### ISSUE TITLE
Update to newer version of Cloudfront API?

### ISSUE THREAD

**Issue by dancoates** [laurilehmijoki/configure-s3-website#16]
There are a couple of issues (laurilehmijoki/s3_website#245, laurilehmijoki/s3_website#240) which I think may be stemming from using a fairly old version of the cloudfront API. There have been [quite a few updates](https://aws.amazon.com/releasenotes/CloudFront) to cloudfront since the 2012-07-01 api version that is used. 

I had a go at swapping out the api version for the latest but it seems that the authentication process the API has changed:

```xml
<?xml version="1.0"?>
<ErrorResponse xmlns="http://cloudfront.amazonaws.com/doc/2016-09-29/">
    <Error>
        <Type>Sender</Type>
        <Code>IncompleteSignature</Code>
        <Message>Authorization header requires 'Credential' parameter. Authorization header requires 'Signature' parameter. Authorization header requires 'SignedHeaders' parameter. </Message>
    </Error>
    <RequestId></RequestId>
</ErrorResponse>
```

**Comment by laurilehmijoki** [comment]
I started the work on upgrading the current CloudFront client to the one included in the official Ruby AWS SDK: https://github.com/laurilehmijoki/configure-s3-website/tree/use-cloudfront-client-from-aws-sdk.

It should solve this problem.

**Comment by dancoates** [comment]
@laurilehmijoki that's brilliant, thank you. Let me know if I can help out with testing at all.

**Comment by laurilehmijoki** [comment]
@dancoates I just force-pushed an update into the dev-branch https://github.com/laurilehmijoki/configure-s3-website/tree/use-cloudfront-client-from-aws-sdk. It seems that when I'm calling [`update_distribution`](https://github.com/laurilehmijoki/configure-s3-website/commit/61345a8db8d776ba6d7e877fc33254b536396c52#diff-4bf2f33b274871649e6fd2b99c2a23e1R26), the Ruby AWS SDK requires me to send more info than what `configure-s3-website` has sent before:

```
~/.rvm/gems/ruby-2.3.0/gems/aws-sdk-core-2.6.13/lib/aws-sdk-core/param_validator.rb:28:in `validate!': parameter validator found 6 errors: (ArgumentError)
  - missing required parameter params[:distribution_config][:origins]
  - missing required parameter params[:distribution_config][:enabled]
  - missing required parameter params[:distribution_config][:default_cache_behavior][:target_origin_id]
  - missing required parameter params[:distribution_config][:default_cache_behavior][:forwarded_values]
  - missing required parameter params[:distribution_config][:default_cache_behavior][:trusted_signers]
  - missing required parameter params[:distribution_config][:default_cache_behavior][:viewer_protocol_policy]
```

I'm planning to fetch the current distribution settings from the CloudFront API and merge them with the ones in the user's local `cloudfront_distribution_config` setting such that the local settings will take precedence. At the moment I don't think any problems will result from this approach. What do you think?

**Comment by dancoates** [comment]
That sounds like the correct approach to me and is in line with the workflow for updating distributions using the aws cli, where you make a call to `get-distribution-config`, make changes to the config and then call `update-distribution`.

**Comment by laurilehmijoki** [comment]
I just force-pushed an implementation that should work the way we've discussed above. 

The unit tests are yet to be written. Then we can release – I'm planning to first put out a release candidate so that @dancoates you can test it before I release it as a part of `s3_website` v3.

This is going to be a backwards incompatible change. I added [migration instructions](https://github.com/laurilehmijoki/configure-s3-website/blob/use-cloudfront-client-from-aws-sdk/changelog.md#breaking-changes) into changelog.md.

**Comment by laurilehmijoki** [comment]
@dancoates I just pushed the version **2.0.0-RC1** into rubygems.org – care to give it a try? Please not that there might be bugs, so it might be a good idea to run it against a test distribution.

You can test it like this:

1. `gem install configure-s3-website --version 2.0.0-RC1`
2. `configure-s3-website --config-file=/tmp/config.yml`
    * test the `config.yml` file can be an `s3_website.yml` file, for example

Please let me know if you need any help with testing. If you deem the release candidate to be ok, I will release it and ship it along with `s3_website` version **3.0.0**.

**Comment by dancoates** [comment]
@laurilehmijoki apologies for the delay in getting back to you on this. I've had some time today to try it out and it seems to work pretty well, though I have certainly not tested comprehensively at this stage.

One thing that I discovered is that the current implementation does not allow updating of hashes within arrays. For example, I'd like to set the `origin_path` property on `distribution_config[:items][0]` but when writing the configuration yaml like:

```
cloudfront_distribution_config:
  origins:
    items:
      - origin_path: /subfolder
```

It doesn't update the existing origin config but instead creates a new item within the origin config with only the `origin_path` property. This results in the following error:

```
parameter validator found 2 errors: (ArgumentError)
  - missing required parameter params[:distribution_config][:origins][:items][1][:id]
  - missing required parameter params[:distribution_config][:origins][:items][1][:domain_name]
```

I also tried specifying the full origin config to add the missing id and domain_name parameters, but that still just prepends the custom origin item config to the items array resulting in:

```
The specified quantity of Origins (1) does not match the actual quantity supplied (2). (Aws::CloudFront::Errors::InconsistentQuantities)
```

I believe that this can be fixed by setting the `:merge_hash_arrays` option on `deep_merge` to true. Unless I am mistaken this is how the configuration merging worked previously?

**Comment by laurilehmijoki** [comment]
@dancoates `2.0.0-RC2` now also merges arrays. I'm using the `:merge_hash_arrays` option, thanks for the tip.

I also updated the release notes on backward incompatibility: https://github.com/laurilehmijoki/configure-s3-website/blob/use-cloudfront-client-from-aws-sdk/changelog.md#breaking-changes

Previously, `configure-s3-website` did not use the `:merge_hash_arrays` option. I guess nobody just tried to do something that required that option.

What are your thoughts on releasing these changes in the version `2.0.0`?

**Comment by dancoates** [comment]
I've spent a good part of today working with the RC1 version with the `:merge_hash_arrays` option added (so the same as what is in RC2). I obviously haven't tested all possible use cases, but my use case of configuring the cloudfront distribution for serving single page apps from a s3 subfolder involves setting quite a few [distribution config options](https://gist.github.com/dancoates/3506d0f9a18a0e6bff797e4632d0d3eb) and I haven't run into any issues with configure-s3-website. The error output from the sdk has also turned out to be very helpful considering how inconsistent the AWS docs are.

So I'd be happy for the changes as they are to be released under `2.0.0` but also understand if you'd like to hold off to do more testing.

**Comment by dancoates** [comment]
Also, just out of curiosity, will you be adding 2.0.0 to s3_website as a new major version?

**Comment by laurilehmijoki** [comment]
Released in version **2.0.0** is now released. Thank you @dancoates for the valuable ideas and testing!

Yes, I will release these changes in the version **3.0.0** of `s3_website`.

**Comment by dancoates** [comment]
no problems @laurilehmijoki thank's for the work on updating.
